### ISSUE TITLE
MultiLayerNetwork serialization problem - rnnTimeStep returns NaN

### ISSUE THREAD

**Issue by jan-stanek** [deeplearning4j/deeplearning4j#2536]
I am using LSTM network from this [example](https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/recurrent/character/GravesLSTMCharModellingExample.java). 

After training I save network to file by ModelSerializer, when I load it, rnnTimeStep returns array full of NaN very often. The longer it is training, the more often it returns NaN. If I don't load network from file, there's no problem with generating.

I am using 0.7.1 version of DL4J and ND4J.

**Comment by AlexDBlack** [comment]
Can you post exactly how you are saving the model, and exactly how you are doing the rnnTimeStep?

**Comment by jan-stanek** [comment]
Save: [https://github.com/jan-stanek/TwitterPredict/blob/master/LstmTrain/src/main/java/Lstm.java#L98](https://github.com/jan-stanek/TwitterPredict/blob/master/LstmTrain/src/main/java/Lstm.java#L98)
rnnTimeStep: [https://github.com/jan-stanek/TwitterPredict/blob/master/Web/src/java/cz/fit/pdd/TwitterPredict/Lstm.java#L28-L72](https://github.com/jan-stanek/TwitterPredict/blob/master/Web/src/java/cz/fit/pdd/TwitterPredict/Lstm.java#L28-L72)

**Comment by AlexDBlack** [comment]
So I can't actually run your code due to the missing pieces :/

But I've adapted the LSTM character example to basically replicate what you are doing with model loader.
You can see it here: https://gist.github.com/AlexDBlack/13d69127bfd7873e443394141c89fb4c

I've run that on both nd4j-native and nd4j-cuda-8.0 (both 0.7.1, both Windows) and I'm not seeing any issues on either.
In both cases, the trained and saved/loaded networks are producing identical outputs.

Can you please run that, and let me know if it is sufficient to reproduce your issue?

**Comment by jan-stanek** [comment]
I've set same parameters and run more epochs without any difference in results.

I'm enclosing simplified code for generating with trained network. [test.zip](https://github.com/deeplearning4j/deeplearning4j/files/665990/test.zip)

It looks like problem appears only when network is trained on GPU, saved and then used on CPU.

**Comment by AlexDBlack** [comment]
OK, thanks. That helps. FWIW I've since run on both native and cuda on linux too, without issue.

I'll try again with that.

**Comment by jan-stanek** [comment]
Thanks. If you want to try training too, you can use LstmTrain module from github with data.txt file from zip.

**Comment by AlexDBlack** [comment]
Confirmed that I'm able to reproduce this with the saved network you provided. CUDA generation seems fine, but native generation is producing NaNs at times.

**Comment by AlexDBlack** [comment]
I'm seeing evidence of exploding activations here - i.e., activations (pre-softmax) are getting very large at some points.

You can step through in a debugger, to confirm:
https://github.com/deeplearning4j/deeplearning4j/blob/7b7d7d89456a935f7b50fe77af325a8edca2313d/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/recurrent/RnnOutputLayer.java#L143
You can inspect the preOutput2d array for the pre-softmax values.

Still not sure if this is the cause, but it does look suspicious...

**Comment by AlexDBlack** [comment]
ok, so we've got this one worked out.
Large activations + nd4j-native backend + bug in softmax 1d edge case -> numerical underflow (that shouldn't be happening)

https://gist.github.com/AlexDBlack/055028d7272f3fbc411b267d460e86ac

Those -4xx values are pre-softmax activations taken from your network (when feeding in all first character), fyi. Shouldn't be an issue normally, just figured I'd mention it (I'd increase l2 regularization a bit, if it was my network).

**Comment by AlexDBlack** [comment]
Fixed here: https://github.com/deeplearning4j/libnd4j/commit/c573f818bc286c6c91df98d6002318738d67ffb6

Thanks for reporting.

**Comment by jan-stanek** [comment]
Thanks
